File: src/main/java/fun/iiii/mixedlogin/LoginServerManager.java
package fun.iiii.mixedlogin;

import com.velocitypowered.api.util.GameProfile;
import fun.iiii.mixedlogin.yggdrasil.VirtualYggdrasilServer;
import fun.iiii.mixedlogin.yggdrasil.offline.VirtualOfflineService;
import fun.iiii.mixedlogin.yggdrasil.offline.VirtualSubService;

import java.util.Optional;

public class LoginServerManager {
    private final VirtualOfflineService virtualOfflineService = new VirtualOfflineService();
    private final VirtualSubService virtualSubService = new VirtualSubService();
    private final VirtualYggdrasilServer offlineYggdrasilServer = new VirtualYggdrasilServer(26748, "127.0.0.1", virtualOfflineService);
    private final VirtualYggdrasilServer subYggdrasilServer = new VirtualYggdrasilServer(26749, "127.0.0.1", virtualSubService);

    public void start() {
        try {
            offlineYggdrasilServer.start();
            subYggdrasilServer.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public boolean shouldOfflineHost(String hostName) {
        if (hostName.isEmpty()) return false;
        if (hostName.startsWith("offline")) return true;
        if (hostName.startsWith("o-")) return true;
        return false;
    }

    public String startOfflineRequest(String userName) {
        return virtualOfflineService.startRequest(userName);
    }


    public void startSubRequest(String serverId, GameProfile gameProfile) {
        virtualSubService.startRequest(serverId,gameProfile);
    }

}

File: openvc/src/main/kotlin/fun/iiii/mixedlogin/LoginServerManager.kt
package `fun`.iiii.mixedlogin

import com.velocitypowered.api.util.GameProfile
import `fun`.iiii.mixedlogin.yggdrasil.VirtualYggdrasilServer
import `fun`.iiii.mixedlogin.yggdrasil.offline.VirtualOfflineService
import `fun`.iiii.mixedlogin.yggdrasil.offline.VirtualSubService

class LoginServerManager {
    private val virtualOfflineService = VirtualOfflineService()
    private val virtualSubService = VirtualSubService()
    private val offlineYggdrasilServer = VirtualYggdrasilServer(
        MixedLoginMain.getConfig().offlineYggdrasil.port,
        MixedLoginMain.getConfig().offlineYggdrasil.host,
        virtualOfflineService
    )
    private val subYggdrasilServer = VirtualYggdrasilServer(
        MixedLoginMain.getConfig().subYggdrasil.port,
        MixedLoginMain.getConfig().subYggdrasil.host,
        virtualSubService
    )

    fun start() {
        try {
            offlineYggdrasilServer.start()
            subYggdrasilServer.start()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun shouldOfflineHost(hostName: String): Boolean {
        if (hostName.isEmpty()) return false
        MixedLoginMain.getConfig().hostMatch.start.forEach {
            if (hostName.startsWith(it)) return true
        }
        return false
    }

    fun startOfflineRequest(userName: String): String {
        return virtualOfflineService.startRequest(userName)
    }

    fun startSubRequest(serverId: String, gameProfile: GameProfile) {
        virtualSubService.startRequest(serverId, gameProfile)
    }
} 